Automated Smoketest: :x:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 063fe1c587d11927e77740e9986ea5effefc0230
Number of Error(s): 6
/mnt/disk1/home/vamosax/smoketest/PR-11266/HPCC-Platform/dali/base/dasess.cpp:2125:16: error: cannot allocate an object of abstract type \xe2\x80\x98CUserDescriptor\xe2\x80\x99

[Build error:](https://gist.github.com/415756ec8457ed9d65469c10fc17814a)
Build log [Build log](https://gist.github.com/HPCCSmoketest/415756ec8457ed9d65469c10fc17814a#file-result-18-05-31-18-50-02-log)

--- a/dali/base/dacoven.cpp
+++ b/dali/base/dacoven.cpp
@@ -37,7 +37,7 @@ extern void closedownDFS();
 // base is saved in store whenever block exhausted, so replacement coven servers can restart 
 
 // server side versioning.
-#define ServerVersion    "3.14"
+#define ServerVersion    "3.15"
 #define MinClientVersion "1.5"
 
 

--- a/dali/base/dasess.cpp
+++ b/dali/base/dasess.cpp
@@ -79,7 +79,7 @@ interface ISessionManagerServer: implements IConnectionMonitor
     virtual void addSession(SessionId id) = 0;
     virtual SessionId lookupProcessSession(INode *node) = 0;
     virtual INode *getProcessSessionNode(SessionId id) =0;
-    virtual SecAccessFlags getPermissionsLDAP(const char *key,const char *obj,IUserDescriptor *udesc,unsigned flags, int *err)=0;
+    virtual SecAccessFlags getPermissionsLDAP(const char *key,const char *obj,IUserDescriptor *udesc,unsigned flags, const char * reqSignature, CDateTime * reqUTCTimestamp, int *err)=0;
     virtual bool clearPermissionsCache(IUserDescriptor *udesc) = 0;
     virtual void stopSession(SessionId sessid,bool failed) = 0;
     virtual void setClientAuth(IDaliClientAuthConnection *authconn) = 0;
@@ -453,6 +453,47 @@ public:
     }
 };
 
+
+//Helper class to Serialize/deserialize digital signature of "key;user;timestamp" and request timestamp
+class CSignatureSerializationHelper
+{
+public:
+    static void serializeSignature(const char * key, IUserDescriptor *udesc, MemoryBuffer &mb)
+    {
+		IDigitalSignatureManager * pDSM = createDigitalSignatureManagerInstanceFromEnv();
+		if (pDSM && pDSM->isDigiSignerConfigured())
+		{
+            //Serialize timestamp and signature of "scope;username;timestamp"
+            //Dali will use this to ensure request came from
+            //valid authenticated user within a reasonable timeframe
+            StringBuffer username;
+            udesc->getUserName(username);
+            CDateTime now;
+            now.setNow();
+            StringBuffer timeStr;
+            now.getString(timeStr, false);//get UTC timestamp
+            VStringBuffer toSign("%s;%s;%s", key, username.str(), timeStr.str());
+
+            StringBuffer b64sig;
+            pDSM->digiSign(toSign, b64sig);//Sign "scope;username;timeStamp"
+
+            //Serialize the signature, and the timestamp object
+            mb.append(b64sig.str());
+            now.serialize(mb);
+        }
+    }
+
+    static void deserializeSignature(MemoryBuffer &mb, StringBuffer & signature, CDateTime & utcTimeStamp)
+    {
+        if (mb.remaining() > 0)
+        {
+            mb.read(signature);
+            utcTimeStamp.deserialize(mb);
+        }
+    }
+};
+
+
 class CSessionRequestServer: public Thread
 {
     bool stopped;
@@ -603,9 +644,14 @@ public:
                 unsigned auditflags = 0;
                 if (mb.length()-mb.getPos()>=sizeof(auditflags))
                     mb.read(auditflags);
-                udesc->deserializeSignature(mb);//deserialize user signature if present
+
+                StringBuffer reqSignature;
+                CDateTime reqUTCTimestamp;
+                if (queryDaliServerVersion().compare("3.15") >= 0)
+                    CSignatureSerializationHelper::deserializeSignature(mb, reqSignature, reqUTCTimestamp);
+
                 int err = 0;
-                SecAccessFlags perms = manager.getPermissionsLDAP(key,obj,udesc,auditflags,&err);
+                SecAccessFlags perms = manager.getPermissionsLDAP(key,obj,udesc,auditflags,reqSignature.str(),&reqUTCTimestamp,&err);
                 mb.clear().append((int)perms);
                 if (err)
                     mb.append(err);
@@ -884,7 +930,7 @@ public:
     }
 
 
-    SecAccessFlags getPermissionsLDAP(const char *key,const char *obj,IUserDescriptor *udesc,unsigned auditflags,int *err)
+    SecAccessFlags getPermissionsLDAP(const char *key,const char *obj,IUserDescriptor *udesc,unsigned auditflags,const char * reqSignature, CDateTime * reqUTCTimestamp, int *err)
     {
         if (err)
             *err = 0;
@@ -912,7 +958,8 @@ public:
 #endif
         udesc->serialize(mb);
         mb.append(auditflags);
-        udesc->serializeSignature(mb);//serialize user signature if Dali version >= 3.14
+        if (queryDaliServerVersion().compare("3.15") >= 0)
+            CSignatureSerializationHelper::serializeSignature(obj, udesc, mb);//serialize signature, timestamp
         if (!queryCoven().sendRecv(mb,RANK_RANDOM,MPTAG_DALI_SESSION_REQUEST,SESSIONREPLYTIMEOUT))
             return SecAccess_None;
         SecAccessFlags perms = SecAccess_Unavailable;
@@ -1130,6 +1177,8 @@ class CLdapWorkItem : public Thread
 {
     StringAttr key;
     StringAttr obj;
+    StringAttr reqSignature;
+    CDateTime  reqUTCTimestamp;
     Linked<IUserDescriptor> udesc;
     Linked<IDaliLdapConnection> ldapconn;
     unsigned flags;
@@ -1144,10 +1193,13 @@ public:
     {
         running = false;
     }
-    void start(const char *_key,const char *_obj,IUserDescriptor *_udesc,unsigned _flags)
+    void start(const char *_key,const char *_obj,IUserDescriptor *_udesc,unsigned _flags,const char * _reqSignature, CDateTime * _reqUTCTimestamp)
     {
         key.set(_key);
         obj.set(_obj); 
+        reqSignature.set(_reqSignature);
+        reqUTCTimestamp.set(*_reqUTCTimestamp);
+
 #ifdef NULL_DALIUSER_STACKTRACE
         StringBuffer sb;
         if (_udesc)
@@ -1175,7 +1227,7 @@ public:
             if (!running)
                 break;
             try {
-                ret = ldapconn->getPermissions(key,obj,udesc,flags);
+                ret = ldapconn->getPermissions(key,obj,udesc,flags,reqSignature.str(),&reqUTCTimestamp);
             }
             catch(IException *e) {
                 LOG(MCoperatorError, unknownJob, e, "CLdapWorkItem"); 
@@ -1405,7 +1457,8 @@ public:
         return NULL;
     }
 
-    virtual SecAccessFlags getPermissionsLDAP(const char *key,const char *obj,IUserDescriptor *udesc,unsigned flags, int *err)
+    //ISessionManagerServer
+    virtual SecAccessFlags getPermissionsLDAP(const char *key,const char *obj,IUserDescriptor *udesc,unsigned flags,const char * reqSignature, CDateTime * reqUTCTimestamp, int *err)
     {
         if (err)
             *err = 0;
@@ -1425,7 +1478,7 @@ public:
         }
 #endif
         if ((ldapconn->getLDAPflags()&(DLF_SAFE|DLF_ENABLED))!=(DLF_SAFE|DLF_ENABLED))
-            return ldapconn->getPermissions(key,obj,udesc,flags);
+            return ldapconn->getPermissions(key,obj,udesc,flags,reqSignature,reqUTCTimestamp);
         atomic_inc(&ldapwaiting);
         unsigned retries = 0;
         while (!stopping) {
@@ -1434,7 +1487,7 @@ public:
                 if (!ldapworker)
                     ldapworker.setown(CLdapWorkItem::get(ldapconn,workthreadsem));
                 if (ldapworker) {
-                    ldapworker->start(key,obj,udesc,flags);
+                    ldapworker->start(key,obj,udesc,flags,reqSignature,reqUTCTimestamp);
                     for (unsigned i=0;i<10;i++) {
                         if (i)
                             WARNLOG("LDAP stalled(%d) - retrying",i);
@@ -1459,7 +1512,7 @@ public:
                             ldapworker.clear();
                             ldapworker.setown(CLdapWorkItem::get(ldapconn,workthreadsem));
                             if (ldapworker)
-                                ldapworker->start(key,obj,udesc,flags);
+                                ldapworker->start(key,obj,udesc,flags,reqSignature,reqUTCTimestamp);
                         }
                     }
                     if (ldapworker)
@@ -2011,12 +2064,7 @@ class CUserDescriptor: implements IUserDescriptor, public CInterface
 {
     StringAttr username;
     StringAttr passwordenc;
-    unsigned sessionToken = 0;//ESP session token
     StringBuffer signature;//user's digital Signature
-    IDigitalSignatureManager * pDSM = nullptr;
-
-    CDateTime utcTimeStamp;//time when request was sent
-    StringBuffer userTimeStampSignature;//signature of "user;timeStamp"
 public:
     IMPLEMENT_IINTERFACE;
     CUserDescriptor()
@@ -2031,21 +2079,9 @@ public:
         decrypt(buf,passwordenc);
         return buf;
     }
-    unsigned querySessionToken()
-    {
-        return sessionToken;
-    }
     const char *querySignature()
     {
         return signature.str();
-    }
-    const CDateTime & queryUTCTimeStamp()
-    {
-        return utcTimeStamp;
-    }
-    const char *queryUserTimeStampSignature()
-    {
-        return userTimeStampSignature.str();
     }
     virtual void set(const char *name,const char *password)
     {
@@ -2057,66 +2093,21 @@ public:
     void set(const char *_name, const char *_password, unsigned _sessionToken, const char *_signature)
     {
         set(_name, _password);
-        sessionToken = _sessionToken;
         signature.clear().append(_signature);
     }
     virtual void clear()
     {
         username.clear();
         passwordenc.clear();
-        sessionToken = 0;
         signature.clear();
-        utcTimeStamp.clear();
-        userTimeStampSignature.clear();
     }
     void serialize(MemoryBuffer &mb)
     {
         mb.append(username).append(passwordenc);
     }
-    void serializeSignature(MemoryBuffer &mb)
-    {
-        if (queryDaliServerVersion().compare("3.14") >= 0)
-        {
-            if (pDSM == nullptr)
-                pDSM = createDigitalSignatureManagerInstanceFromEnv();
-
-            //Serialize timestamp and signature of "username;timestamp"
-            //Dali will use this to ensure request came from
-            //valid authenticated user within a reasonable timeframe
-            if (pDSM && pDSM->isDigiSignerConfigured())
-            {
-                CDateTime now;
-                now.setNow();
-                StringBuffer timeStr;
-                now.getString(timeStr, false);//get UTC timestamp
-                VStringBuffer toSign("%s;%s", username.str(), timeStr.str());
-
-                StringBuffer userTimeStampSignature;
-                pDSM->digiSign(toSign, userTimeStampSignature);//Sign "username;timeStamp"
-
-                //Serialize the timestamped signature, and the timestamp object
-                mb.append(userTimeStampSignature.str());
-                now.serialize(mb);
-            }
-        }
-    }
     void deserialize(MemoryBuffer &mb)
     {
         mb.read(username).read(passwordenc);
-    }
-    void deserializeSignature(MemoryBuffer &mb)
-    {
-        utcTimeStamp.clear();
-        userTimeStampSignature.clear();
-        if (queryDaliServerVersion().compare("3.14") >= 0)
-        {
-            if (mb.remaining() > 0)
-            {
-                mb.read(userTimeStampSignature);
-                if (mb.remaining() > 0)
-                    utcTimeStamp.deserialize(mb);
-            }
-        }
     }
 };
 

--- a/dali/base/dasess.cpp
+++ b/dali/base/dasess.cpp
@@ -454,11 +454,11 @@ public:
 };
 
 
-//Helper class to Serialize/deserialize digital signature of "key;user;timestamp" and request timestamp
-class CSignatureSerializationHelper
+//Helper class to Serialize/deserialize digital signature of "scope;username;timestamp" and request timestamp
+class CDaliMessageSignatureHelper
 {
 public:
-    static void serializeSignature(const char * key, IUserDescriptor *udesc, MemoryBuffer &mb)
+    static void serializeSignature(const char * scope, IUserDescriptor *udesc, MemoryBuffer &mb)
     {
 		IDigitalSignatureManager * pDSM = createDigitalSignatureManagerInstanceFromEnv();
 		if (pDSM && pDSM->isDigiSignerConfigured())
@@ -472,7 +472,7 @@ public:
             now.setNow();
             StringBuffer timeStr;
             now.getString(timeStr, false);//get UTC timestamp
-            VStringBuffer toSign("%s;%s;%s", key, username.str(), timeStr.str());
+            VStringBuffer toSign("%s;%s;%s", scope, username.str(), timeStr.str());
 
             StringBuffer b64sig;
             pDSM->digiSign(toSign, b64sig);//Sign "scope;username;timeStamp"
@@ -648,7 +648,7 @@ public:
                 StringBuffer reqSignature;
                 CDateTime reqUTCTimestamp;
                 if (queryDaliServerVersion().compare("3.15") >= 0)
-                    CSignatureSerializationHelper::deserializeSignature(mb, reqSignature, reqUTCTimestamp);
+                    CDaliMessageSignatureHelper::deserializeSignature(mb, reqSignature, reqUTCTimestamp);
 
                 int err = 0;
                 SecAccessFlags perms = manager.getPermissionsLDAP(key,obj,udesc,auditflags,reqSignature.str(),&reqUTCTimestamp,&err);
@@ -959,7 +959,7 @@ public:
         udesc->serialize(mb);
         mb.append(auditflags);
         if (queryDaliServerVersion().compare("3.15") >= 0)
-            CSignatureSerializationHelper::serializeSignature(obj, udesc, mb);//serialize signature, timestamp
+            CDaliMessageSignatureHelper::serializeSignature(obj, udesc, mb);//serialize scope, signature, timestamp
         if (!queryCoven().sendRecv(mb,RANK_RANDOM,MPTAG_DALI_SESSION_REQUEST,SESSIONREPLYTIMEOUT))
             return SecAccess_None;
         SecAccessFlags perms = SecAccess_Unavailable;

--- a/dali/base/dasess.cpp
+++ b/dali/base/dasess.cpp
@@ -603,7 +603,7 @@ public:
                 unsigned auditflags = 0;
                 if (mb.length()-mb.getPos()>=sizeof(auditflags))
                     mb.read(auditflags);
-                udesc->deserializeExtra(mb);//deserialize user signature if present
+                udesc->deserializeSignature(mb);//deserialize user signature if present
                 int err = 0;
                 SecAccessFlags perms = manager.getPermissionsLDAP(key,obj,udesc,auditflags,&err);
                 mb.clear().append((int)perms);
@@ -912,7 +912,7 @@ public:
 #endif
         udesc->serialize(mb);
         mb.append(auditflags);
-        udesc->serializeExtra(mb);//serialize user signature if Dali version >= 3.14
+        udesc->serializeSignature(mb);//serialize user signature if Dali version >= 3.14
         if (!queryCoven().sendRecv(mb,RANK_RANDOM,MPTAG_DALI_SESSION_REQUEST,SESSIONREPLYTIMEOUT))
             return SecAccess_None;
         SecAccessFlags perms = SecAccess_Unavailable;
@@ -2014,6 +2014,9 @@ class CUserDescriptor: implements IUserDescriptor, public CInterface
     unsigned sessionToken = 0;//ESP session token
     StringBuffer signature;//user's digital Signature
     IDigitalSignatureManager * pDSM = nullptr;
+
+    CDateTime utcTimeStamp;//time when request was sent
+    StringBuffer userTimeStampSignature;//signature of "user;timeStamp"
 public:
     IMPLEMENT_IINTERFACE;
     CUserDescriptor()
@@ -2036,6 +2039,14 @@ public:
     {
         return signature.str();
     }
+    const CDateTime & queryUTCTimeStamp()
+    {
+        return utcTimeStamp;
+    }
+    const char *queryUserTimeStampSignature()
+    {
+        return userTimeStampSignature.str();
+    }
     virtual void set(const char *name,const char *password)
     {
         username.set(name);
@@ -2055,36 +2066,56 @@ public:
         passwordenc.clear();
         sessionToken = 0;
         signature.clear();
+        utcTimeStamp.clear();
+        userTimeStampSignature.clear();
     }
     void serialize(MemoryBuffer &mb)
     {
         mb.append(username).append(passwordenc);
     }
-    void serializeExtra(MemoryBuffer &mb)
+    void serializeSignature(MemoryBuffer &mb)
     {
         if (queryDaliServerVersion().compare("3.14") >= 0)
         {
-            if (signature.isEmpty())
+            if (pDSM == nullptr)
+                pDSM = createDigitalSignatureManagerInstanceFromEnv();
+
+            //Serialize timestamp and signature of "username;timestamp"
+            //Dali will use this to ensure request came from
+            //valid authenticated user within a reasonable timeframe
+            if (pDSM && pDSM->isDigiSignerConfigured())
             {
-                if (pDSM == nullptr)
-                    pDSM = createDigitalSignatureManagerInstanceFromEnv();
-                if (pDSM && pDSM->isDigiSignerConfigured())
-                    pDSM->digiSign(username, signature);//Set user's digital signature
+                CDateTime now;
+                now.setNow();
+                StringBuffer timeStr;
+                now.getString(timeStr, false);//get UTC timestamp
+                VStringBuffer toSign("%s;%s", username.str(), timeStr.str());
+
+                StringBuffer userTimeStampSignature;
+                pDSM->digiSign(toSign, userTimeStampSignature);//Sign "username;timeStamp"
+
+                //Serialize the timestamped signature, and the timestamp object
+                mb.append(userTimeStampSignature.str());
+                now.serialize(mb);
             }
-            mb.append(signature);
         }
     }
     void deserialize(MemoryBuffer &mb)
     {
         mb.read(username).read(passwordenc);
     }
-    void deserializeExtra(MemoryBuffer &mb)
+    void deserializeSignature(MemoryBuffer &mb)
     {
-        signature.clear();
+        utcTimeStamp.clear();
+        userTimeStampSignature.clear();
         if (queryDaliServerVersion().compare("3.14") >= 0)
         {
             if (mb.remaining() > 0)
-                mb.read(signature);
+            {
+                mb.read(userTimeStampSignature);
+                if (mb.remaining() > 0)
+                    utcTimeStamp.deserialize(mb);
+            }
         }
     }
 };

--- a/dali/base/dasess.hpp
+++ b/dali/base/dasess.hpp
@@ -77,14 +77,9 @@ interface IUserDescriptor: extends serializable
     virtual StringBuffer &getUserName(StringBuffer &buf)=0;
     virtual StringBuffer &getPassword(StringBuffer &buf)=0;
     virtual const char *querySignature()=0;//user's digital signature
-    virtual unsigned querySessionToken()=0;//ESP session token
-    virtual const CDateTime & queryUTCTimeStamp()=0;//Time stamp when a Dali request was requested
-    virtual const char *queryUserTimeStampSignature()=0;//digital signature of "username;TimeStampString"
     virtual void set(const char *name,const char *password)=0;
     virtual void set(const char *name,const char *password, unsigned sessionToken, const char *_signature)=0;
     virtual void clear()=0;
-    virtual void serializeSignature(MemoryBuffer &tgt)=0;
-    virtual void deserializeSignature(MemoryBuffer &src)=0;
 };
 
 extern da_decl IUserDescriptor *createUserDescriptor();
@@ -119,7 +114,7 @@ interface ISessionManager: extends IInterface
     virtual StringBuffer &getClientProcessEndpoint(SessionId id,StringBuffer &buf)=0; // for diagnostics
     virtual unsigned queryClientCount() = 0; // for SNMP
 
-    virtual SecAccessFlags getPermissionsLDAP(const char *key,const char *obj,IUserDescriptor *udesc,unsigned auditflags, int *err=NULL)=0;
+    virtual SecAccessFlags getPermissionsLDAP(const char *key,const char *obj,IUserDescriptor *udesc,unsigned auditflags, const char * reqSignature=nullptr, CDateTime * reqUTCTimestamp=nullptr, int *err=NULL)=0;
     virtual bool checkScopeScansLDAP()=0;
     virtual unsigned getLDAPflags()=0;
     virtual void setLDAPflags(unsigned flags)=0;

--- a/dali/base/dasess.hpp
+++ b/dali/base/dasess.hpp
@@ -27,6 +27,7 @@
 #endif
 
 #include "dacoven.hpp"
+#include "jtime.hpp"
 
 typedef DALI_UID SessionId;
 typedef DALI_UID SubscriptionId;
@@ -77,11 +78,13 @@ interface IUserDescriptor: extends serializable
     virtual StringBuffer &getPassword(StringBuffer &buf)=0;
     virtual const char *querySignature()=0;//user's digital signature
     virtual unsigned querySessionToken()=0;//ESP session token
+    virtual const CDateTime & queryUTCTimeStamp()=0;//Time stamp when a Dali request was requested
+    virtual const char *queryUserTimeStampSignature()=0;//digital signature of "username;TimeStampString"
     virtual void set(const char *name,const char *password)=0;
     virtual void set(const char *name,const char *password, unsigned sessionToken, const char *_signature)=0;
     virtual void clear()=0;
-    virtual void serializeExtra(MemoryBuffer &tgt)=0;
-    virtual void deserializeExtra(MemoryBuffer &src)=0;
+    virtual void serializeSignature(MemoryBuffer &tgt)=0;
+    virtual void deserializeSignature(MemoryBuffer &src)=0;
 };
 
 extern da_decl IUserDescriptor *createUserDescriptor();

--- a/dali/server/daldap.cpp
+++ b/dali/server/daldap.cpp
@@ -121,7 +121,7 @@ public:
     }
 
 
-    SecAccessFlags getPermissions(const char *key,const char *obj,IUserDescriptor *udesc,unsigned auditflags)
+    SecAccessFlags getPermissions(const char *key,const char *obj,IUserDescriptor *udesc,unsigned auditflags,const char * reqSignature, CDateTime * reqUTCTimestamp)
     {
         if (!ldapsecurity||((getLDAPflags()&DLF_ENABLED)==0)) 
             return SecAccess_Full;
@@ -149,19 +149,19 @@ public:
         bool authenticated = false;
 
         //Check that the digital signature provided by the caller (signature of
-        //caller's "username;timeStamp") matches what we expect it to be
-        if (!isEmptyString(udesc->queryUserTimeStampSignature()))
+        //caller's "scope;username;timeStamp") matches what we expect it to be
+        if (!isEmptyString(reqSignature))
         {
             if (nullptr == pDSM)
                 pDSM = createDigitalSignatureManagerInstanceFromEnv();
             if (pDSM && pDSM->isDigiVerifierConfigured())
             {
                 StringBuffer requestTimestamp;
-                udesc->queryUTCTimeStamp().getString(requestTimestamp, false);//extract timestamp string from Dali request
+                reqUTCTimestamp->getString(requestTimestamp, false);//extract timestamp string from Dali request
 
                 CDateTime now;
                 now.setNow();
-                if (now.compare(udesc->queryUTCTimeStamp()) < 0)//timestamp from the future?
+                if (now.compare(*reqUTCTimestamp) < 0)//timestamp from the future?
                 {
                     ERRLOG("LDAP: getPermissions(%s) scope=%s user=%s Request digital signature timestamp %s from the future",key?key:"NULL",obj?obj:"NULL",username.str(), requestTimestamp.str());
                     return SecAccess_None;//deny
@@ -171,16 +171,16 @@ public:
                 expiry.set(now);
                 expiry.adjustTime(requestSignatureExpiryMinutes);//compute expiration timestamp
 
-                if (expiry.compare(udesc->queryUTCTimeStamp()) < 0)//timestamp too far in the past?
+                if (expiry.compare(*reqUTCTimestamp) < 0)//timestamp too far in the past?
                 {
                     ERRLOG("LDAP: getPermissions(%s) scope=%s user=%s Expired request digital signature timestamp %s",key?key:"NULL",obj?obj:"NULL",username.str(), requestTimestamp.str());
                     return SecAccess_None;//deny
                 }
 
-                VStringBuffer expectedUserDateStr("%s;%s", username.str(), requestTimestamp.str());
-                StringBuffer b64Signature(udesc->queryUserTimeStampSignature());//extract user;timestamp signature from Dali request
+                VStringBuffer expectedStr("%s;%s;%s", obj, username.str(), requestTimestamp.str());
+                StringBuffer b64Signature(reqSignature);// signature of scope;user;timestamp
 
-                if (!pDSM->digiVerify(expectedUserDateStr, b64Signature))//digital signature match what we expect?
+                if (!pDSM->digiVerify(expectedStr, b64Signature))//does the digital signature match what we expect?
                 {
                     ERRLOG("LDAP: getPermissions(%s) scope=%s user=%s fails digital signature verification",key?key:"NULL",obj?obj:"NULL",username.str());
                     return SecAccess_None;//deny

--- a/dali/server/daldap.cpp
+++ b/dali/server/daldap.cpp
@@ -50,6 +50,7 @@ class CDaliLdapConnection: implements IDaliLdapConnection, public CInterface
     StringAttr              filesdefaultuser;
     StringAttr              filesdefaultpassword;
     unsigned                ldapflags;
+    unsigned                requestSignatureExpiryMinutes;//Age at which a dali permissions request signature becomes invalid
     IDigitalSignatureManager * pDSM = nullptr;
 
     void createDefaultScopes()
@@ -81,6 +82,7 @@ public:
     {
         ldapflags = 0;
         if (ldapprops) {
+            requestSignatureExpiryMinutes = ldapprops->getPropInt("@reqSignatureExpiry", 10);
             if (ldapprops->getPropBool("@checkScopeScans",true))
                 ldapflags |= DLF_SCOPESCANS;
             if (ldapprops->getPropBool("@safeLookup",true))
@@ -144,24 +146,50 @@ public:
         Owned<ISecUser> user = ldapsecurity->createUser(username);
         user->credentials().setPassword(password);
 
-        //Check user's digital signature, if present
         bool authenticated = false;
-        if (!isEmptyString(udesc->querySignature()))
+
+        //Check that the digital signature provided by the caller (signature of
+        //caller's "username;timeStamp") matches what we expect it to be
+        if (!isEmptyString(udesc->queryUserTimeStampSignature()))
         {
             if (nullptr == pDSM)
                 pDSM = createDigitalSignatureManagerInstanceFromEnv();
             if (pDSM && pDSM->isDigiVerifierConfigured())
             {
-                StringBuffer b64Signature(udesc->querySignature());
-                if (!pDSM->digiVerify(username, b64Signature))//digital signature valid?
+                StringBuffer requestTimestamp;
+                udesc->queryUTCTimeStamp().getString(requestTimestamp, false);//extract timestamp string from Dali request
+
+                CDateTime now;
+                now.setNow();
+                if (now.compare(udesc->queryUTCTimeStamp()) < 0)//timestamp from the future?
                 {
-                    ERRLOG("LDAP: getPermissions(%s) scope=%s user=%s invalid user digital signature",key?key:"NULL",obj?obj:"NULL",username.str());
+                    ERRLOG("LDAP: getPermissions(%s) scope=%s user=%s Request digital signature timestamp %s from the future",key?key:"NULL",obj?obj:"NULL",username.str(), requestTimestamp.str());
                     return SecAccess_None;//deny
                 }
-                else
-                    authenticated = true;
+
+                CDateTime expiry;
+                expiry.set(now);
+                expiry.adjustTime(requestSignatureExpiryMinutes);//compute expiration timestamp
+
+                if (expiry.compare(udesc->queryUTCTimeStamp()) < 0)//timestamp too far in the past?
+                {
+                    ERRLOG("LDAP: getPermissions(%s) scope=%s user=%s Expired request digital signature timestamp %s",key?key:"NULL",obj?obj:"NULL",username.str(), requestTimestamp.str());
+                    return SecAccess_None;//deny
+                }
+
+                VStringBuffer expectedUserDateStr("%s;%s", username.str(), requestTimestamp.str());
+                StringBuffer b64Signature(udesc->queryUserTimeStampSignature());//extract user;timestamp signature from Dali request
+
+                if (!pDSM->digiVerify(expectedUserDateStr, b64Signature))//digital signature match what we expect?
+                {
+                    ERRLOG("LDAP: getPermissions(%s) scope=%s user=%s fails digital signature verification",key?key:"NULL",obj?obj:"NULL",username.str());
+                    return SecAccess_None;//deny
+                }
+
+                authenticated = true;//Digital signature verified
             }
-            user->credentials().setSignature(udesc->querySignature());
+            else
+                ERRLOG("LDAP: getPermissions(%s) scope=%s user=%s digital signature support not available",key?key:"NULL",obj?obj:"NULL",username.str());
         }
 
         if (!authenticated && !ldapsecurity->authenticateUser(*user, NULL))

--- a/dali/server/daldap.hpp
+++ b/dali/server/daldap.hpp
@@ -26,7 +26,7 @@ interface IUserDescriptor;
 
 interface IDaliLdapConnection: extends IInterface
 {
-    virtual SecAccessFlags getPermissions(const char *key,const char *obj,IUserDescriptor *udesc,unsigned auditflags)=0;
+    virtual SecAccessFlags getPermissions(const char *key,const char *obj,IUserDescriptor *udesc,unsigned auditflags,const char * reqSignature, CDateTime * reqUTCTimestamp)=0;
     virtual bool checkScopeScans() = 0;
     virtual unsigned getLDAPflags() = 0;
     virtual void setLDAPflags(unsigned flags) = 0;

--- a/initfiles/componentfiles/config2xml/dali.xsd
+++ b/initfiles/componentfiles/config2xml/dali.xsd
@@ -419,6 +419,13 @@ Add a key for dali server names named dali_name that can be referred to from els
                 </xs:appinfo>
             </xs:annotation>
         </xs:attribute>
+        <xs:attribute name="reqSignatureExpiry" use="optional" type="xs:string" default="10">
+            <xs:annotation>
+                <xs:appinfo>
+                    <tooltip>Lifetime in minutes of a permissions request digital signature.</tooltip>
+                </xs:appinfo>
+            </xs:annotation>
+        </xs:attribute>
         <xs:attribute name="checkScopeScans" type="xs:boolean" use="optional" default="true">
             <xs:annotation>
                 <xs:appinfo>

--- a/initfiles/componentfiles/configxml/dali.xsd
+++ b/initfiles/componentfiles/configxml/dali.xsd
@@ -402,6 +402,13 @@
         </xs:appinfo>
       </xs:annotation>
     </xs:attribute>
+    <xs:attribute name="reqSignatureExpiry" use="optional" type="xs:string" default="10">
+      <xs:annotation>
+        <xs:appinfo>
+          <tooltip>Lifetime in minutes of a permissions request digital signature.</tooltip>
+        </xs:appinfo>
+      </xs:annotation>
+    </xs:attribute>
     <xs:attribute name="checkScopeScans" type="xs:boolean" use="optional" default="true">
       <xs:annotation>
         <xs:appinfo>

--- a/initfiles/componentfiles/configxml/dali.xsl
+++ b/initfiles/componentfiles/configxml/dali.xsl
@@ -240,7 +240,7 @@
         </xsl:element>
         <xsl:if test="string(@ldapServer) != ''">
           <xsl:element name="ldapSecurity">
-            <xsl:copy-of select="@ldapProtocol | @authMethod | @maxConnections | @workunitsBasedn | @filesDefaultUser | @filesDefaultPassword"/>
+            <xsl:copy-of select="@ldapProtocol | @authMethod | @maxConnections | @workunitsBasedn | @filesDefaultUser | @filesDefaultPassword | @reqSignatureExpiry"/>
             <xsl:variable name="ldapServerName" select="@ldapServer"/>
             <xsl:attribute name="filesBasedn">
                 <xsl:value-of select="/Environment/Software/LDAPServerProcess[@name=$ldapServerName]/@filesBasedn"/>

